### DISCUSSION TITLE
Execute `ipnb` File via API, resp. in Batch mode

### DISCUSSION THREAD

**Discussion by FlorianKummer** [dotnet/interactive#1390] [Q&A]
For our research code, in the past, we used some C#-scripting module from Mono (`Mono.CSharp.dll`) as a scripting frontend, we even created our own, custom (but glitchy) notebook GUI. 
Now, we want to move everything towards dotnet-interactive, resp. Jupyter notebooks.

A typical use-case is the following: 
We further used these worksheets in our test-procedure:
* We developed and implemented some numerical method, 
* wrote a notebook to do some calculations, 
* wrote a scientific publication and then 
* added the notebook to our set of tests in order to ensure that certain numerical properties still hold when the code is developed.

I allready managed to convert our custom notebooks to `ipnb`, using mostly members of the `Microsoft.DotNet.Interactive.Notebook` namespace.

Now, I also want to execute those notebooks which we use for testing as a unit test. I could theoretically use `jupyter nbconvert --execute ...`, but I'm wondering if there is any option which does not depend on an existing Jupyter installation.

I already managed to execute a single command by `CSharpKerner.SendAsync(...)`, which seems to work; at least, I can fetch variables through `CSharpKernel.TryGetVariable(...)`. However, I am missing some general way to check if the execution went well or not, i.e. a check for uncaught exceptions, syntax errors, etc.

Is there any way of executing an entire notebook and check whether some Exception or NUnit Assertion has been raised, or if any Syntax error occurs?
(Either command-by-command or the entire notebook at once?)

**Comment by jonsequitur** [comment]
This is something we're working on. You can see the related issues under this label: https://github.com/dotnet/interactive/issues?q=is%3Aopen+is%3Aissue+label%3AArea-Automation

The events returned by the `KernelCommandResult.KernelEvents` observable will tell you if something went wrong. The sequence will conclude with either a `CommandSucceeded` or `CommandFailed` event. Exceptions from the notebook code will be wrapped into the `CommandFailed` event.

**Comment by FlorianKummer** [comment]
Thanks, this already helps me a bit futher;

Which kernel would you recommend for this kind of 'end-user'-interaction?
I tried the following:
```
   CSharpKernel csk = new CSharpKernel();
   var obs = new MeObservator(); // implements IObserver<KernelEvent>
   csk.KernelEvents.Subscribe(obs);
   
   var t4 = csk.SendAsync(new SubmitCode("sdjhsjkahdk  error!"), CancellationToken.None);
   t4.Wait();
```
However, this seems to cause some sort of infinite loop, `MeObservator.OnNext(...)` is beeing called over and over and the task never completes.

**Reply by jonsequitur** [reply]
The `Kernel.KernelEvents` observable doesn't terminate. If you want an observable scoped just to the command you sent, you can do this:

```csharp
var result = await csk.SendAsync(new SubmitCode("sdjhsjkahdk  error!"), CancellationToken.None);
result.KernelEvents.Subscribe(obs);
```

Note that in this model, you won't actually receive any events until they've all been emitted. The `KernelCommandResult.KernelEvents` observable is a `ReplayObservable`. 

Another approach, if real-time updates are important and processing the command might take a while, is to subscribe to the `Kernel.KernelEvents` only until you receive a completion event for the command, i.e. a `CommandSucceeded` or a `CommandFailed`.
